Content.Shared/Paper/PaperSystem.cs

```diff
@@ -265,6 +271,85 @@ public bool TryStamp(Entity<PaperComponent> entity, StampDisplayInfo stampInfo,
         return true;
     }
 
+        // Umbra - Begin - Paper signing
+        // Send paper signing alt verb to the client if applicable.
+        // Based on LockSystem.cs for alt-click behavior.
+        private void AddSignVerb(Entity<PaperComponent> ent, ref GetVerbsEvent<AlternativeVerb> args)
+        {
+            if (!args.CanAccess || !args.CanInteract)
+                return;
+
+            // Sanity check
+            if (ent.Owner != args.Target)
+                return;
+
+            // Pens have a `Write` tag.
+            if (args.Using is not { } item|| !_tagSystem.HasTag(item, WriteTag))
+                return;
+
+            // If theres a fake signature available, get it
+            TryComp<ForgeSignatureComponent>(item, out var fakeSignatureComp);
+
+            var user = args.User;
+
+            AlternativeVerb verb = new()
+            {
+                Act = () => TrySign(ent, user, fakeSignatureComp),
+                Text = Loc.GetString("paper-component-verb-sign"),
+                // Icon = Don't have an icon yet. Todo for later.
+            };
+            args.Verbs.Add(verb);
+        }
+
+        // Umbra: Actual signature code.
+        private bool TrySign(Entity<PaperComponent> ent, EntityUid signer, ForgeSignatureComponent? signatureComp)
+        {
+            var signature = (signatureComp == null || signatureComp.Signature == "" ? Name(signer) : signatureComp.Signature);
+
+            // Generate display information.
+            var info = new StampDisplayInfo
+            {
+                StampedName = signature,
+                StampedColor = SignatureColor,
+                Type = StampType.Signature,
+            };
+
+            // Try stamp with the info, return false if failed.
+            if (!TryStamp(ent, info, "paper_stamp-generic"))
+                return false;
+
+            // Signing successful, popup time.
+
+            _popupSystem.PopupEntity(
+                Loc.GetString(
+                    "paper-component-action-signed-other",
+                    ("user", signer),
+                    ("target", ent)
+                ),
+                signer,
+                Filter.PvsExcept(signer, entityManager: EntityManager),
+                true
+            );
+
+            _popupSystem.PopupEntity(
+                Loc.GetString(
+                    "paper-component-action-signed-self",
+                    ("target", ent)
+                ),
+                signer,
+                signer
+            );
```

Comment:
```suggestion
        _popupSystem.PopupPredicted(
            Loc.GetString(
                "paper-component-action-signed-self",
                ("target", ent)
            ),
            Loc.GetString(
                "paper-component-action-signed-other",
                ("user", signer),
                ("target", ent)
            ),
            signer,
            signer
        );
```
(indentation's gonna be wrong since I'm copy-pasting out of Rider which fixed the indentation)

![Image](https://github.com/user-attachments/assets/c98b9246-fd2d-4aaf-a37e-7935bd09155c)